### ISSUE TITLE
Set `file_temporary_path` in initial `system.core.json` file

### ISSUE THREAD

**Issue by alanmels** [backdrop/backdrop-issues#4193]
**Description of the bug**

Despite the `/admin/config/media/file-system` page on a freshly installed Backdrop site shows that **Temporary directory** is set to `/tmp`:

<img width="824" alt="Screen Shot 2019-11-03 at 3 46 43 AM" src="https://user-images.githubusercontent.com/56604405/68078706-da6a0080-fdec-11e9-8932-6fb556090bda.png">

the `file_temporary_path` variable in the `system.core.json` file in active configuration directory shows empty. Therefore the function `config_get('system.core', 'file_temporary_path')` returns empty value causing number of serious troubles.

**Steps To Reproduce**

Just deploy a new Backdrop site and see the described difference between UI and configuration file.

You can quickly fire up the `cat files/config_*/active/system.core.json | grep file_temporary_path` command on CLI to see:

```
    "file_temporary_path": "",
```

At the same time, if you go to the `/admin/config/media/file-system` page then it shows `/tmp` for Temporary directory.

**Expected behavior**

The `file_temporary_path` variable in `system.core.json` file should match what the website's user interface shows for **Temporary directory**.

**Additional information**

If you save the `/admin/config/media/file-system` page, then the `file_temporary_path` variable in `system.core.json` file gets correct value. However, we should not expect that every user necessarily goes to and saves the `/admin/config/media/file-system` page just after installation. 

To me this seems pretty urgent issue to fix, so I'll be submitting PR soon.

---
~~PR by @alanmels: https://github.com/backdrop/backdrop/pull/2971~~
~~New PR by @alanmels: https://github.com/backdrop/backdrop/pull/2972~~
Yet another PR by @alanmels: https://github.com/backdrop/backdrop/pull/2988

**Comment by herbdool** [comment]
I suspect this isn't a bug. If you are looking for the temp directory in a custom module or something you might need to use `file_directory_temp`. I haven't tested this.

The function makes some guesses of the temp directory if it's not set in config.

**Comment by alanmels** [comment]
I've just tested this on a remote server with different OS and the **Temporary directory** showed a directory different than `/tmp`, so my PR is not good as I assumed the system always use `/tmp` initially.

At the same time, the `config_get('system.core', 'file_temporary_path')` still returns empty despite UI tells the directory is set.

**Comment by BWPanda** [comment]
The issue is that `file_directory_temp()` will find and use a different tmp directory if the config value is empty. This is good, but I think we need to make it save the new, found tmp directory back to config so that there's no inconsistency here.

**Comment by alanmels** [comment]
Drupal 8 seems to start with empty temp directory on both UI and ` \Drupal::config('system.file')->get('path.temporary');`, so they do not have contradiction. On the contrary Backrop shows a path to the directory on UI from the start, but misses the actual path in config file.

**Comment by alanmels** [comment]
I've closed my PR as the issue must be addressed differently.

**Comment by BWPanda** [comment]
Looks good to me!

**Comment by indigoxela** [comment]
Does it need to be set anyway? _PHP has default values for that._

Grabbed from a php.ini:
```
; Directory where the temporary files should be placed.
; Defaults to the system default (see sys_get_temp_dir)
; sys_temp_dir = "/tmp"
```
```
; Temporary directory for HTTP uploaded files (will use system default if not
; specified).
; http://php.net/upload-tmp-dir
;upload_tmp_dir =
```
Setting it might eventually cause problems when a site is moved/copied/pushed to the prod server.

**Comment by indigoxela** [comment]
For the curious and for further reading:

class BackdropTemporaryStreamWrapper: https://github.com/backdrop/backdrop/blob/faead38e3753b703919850b7444954f20b708365/core/includes/stream_wrappers.inc#L965

function file_directory_temp: https://api.backdropcms.org/api/backdrop/core%21includes%21file.inc/function/file_directory_temp/1

**Comment by BWPanda** [comment]
In my opinion, Backdrop already has a way to set the temporary directory, but there's an inconsistency between the value used and the value displayed. That's a bug, and that's what this issue is trying to fix.

Discussion about whether Backdrop should have a setting for this in the first place, to me, seems like it should be a separate issue...

**Comment by alanmels** [comment]
@indigoxela, thanks for the input. As @BWPanda said there is inconsistency between what the core sets the temporary directory and what user sees on Backdrop UI.

**Comment by jenlampton** [comment]
> The issue is that file_directory_temp() will find and use a different tmp directory if the config value is empty. This is good, but I think we need to make it save the new, found tmp directory back to config so that there's no inconsistency here.

I really like @BWPanda's idea here. If there is some magic happening behind the scenes to determine a temp directory, the result of that magic needs to be visible to the end-user. 
I'd prefer that Backdrop determine the location of `tmp`, set the config value, and then everything else from that point on use what is defined in config. 

We might also want to add some description text to the temp field in the UI saying that Backdrop has automatically determined the location of the temporary directory. 

We might also want to add a feature: if the value is left empty, Backdrop goes and gets the system tmp directory again, and saves it again. This would give us an easy way to reset for changing servers). Maybe this last bit is a separate issue...

**Comment by alanmels** [comment]
> I'd prefer that Backdrop determine the location of `tmp`

Yep, Backdrop does that in the following function:

```
/**
 * Gets the path of system-appropriate temporary directory.
 */
function file_directory_temp() {
  $temporary_directory = config_get('system.core', 'file_temporary_path');
  if (empty($temporary_directory)) {
    $directories = array();
    // Has PHP been set with an upload_tmp_dir?
    if (ini_get('upload_tmp_dir')) {
      $directories[] = ini_get('upload_tmp_dir');
    }
    // Operating system specific dirs.
    if (substr(PHP_OS, 0, 3) == 'WIN') {
      $directories[] = 'c:\\windows\\temp';
      $directories[] = 'c:\\winnt\\temp';
    }
    else {
      $directories[] = '/tmp';
    }
    // PHP may be able to find an alternative tmp directory.
    $directories[] = sys_get_temp_dir();
    foreach ($directories as $directory) {
      if (is_dir($directory) && is_writable($directory)) {
        $temporary_directory = $directory;
        break;
      }
    }
    if (empty($temporary_directory)) {
      // If no directory has been found default to 'files/tmp'.
      $temporary_directory = config_get('system.core', 'file_public_path') . '/tmp';
      // Windows accepts paths with either slash (/) or backslash (\), but will
      // not accept a path which contains both a slash and a backslash. Since
      // the 'file_public_path' variable may have either format, we sanitize
      // everything to use slash which is supported on all platforms.
      $temporary_directory = str_replace('\\', '/', $temporary_directory);
    }
  }
```

>  set the config value, and then everything else from that point on use what is defined in config.

The above function was missing the below code which saves the config value:

```
 // Now when the temporary directory is found, we need to save it to
    // the configuration file.
    config_set('system.core', 'file_temporary_path', $temporary_directory);
```

> We might also want to add some description text to the temp field in the UI saying that Backdrop has automatically determined the location of the temporary directory.

I didn't touch UI, but if you want I can add such a text.
 
> We might also want to add a feature: if the value is left empty, Backdrop goes and gets the system tmp directory again, and saves it again. (This would give us an easy way to "reset" the value when changing servers). Maybe this last bit is a separate issue...

I believe the above function does exactly that. Please see the commit page: https://github.com/alanmels/backdrop/commit/feffc3774baf92ee18da1f3ca92f28f43353ffdd

**Comment by jenlampton** [comment]
When is `file_directory_temp()` called? I'm guessing when the form is saved, and install? Any other times?

**Comment by alanmels** [comment]
> When is `file_directory_temp()` called? I'm guessing when the form is saved, and on install? Any other time?

According to https://api.backdropcms.org/api/backdrop/core%21includes%21file.inc/function/calls/file_directory_temp/1 there are six calls to the function:


1. BackdropTemporaryStreamWrapper::getDirectoryPath in core/includes/stream_wrappers.inc
Implements abstract public function getDirectoryPath()
2. ConfigurationUITest::testExport in core/modules/config/tests/config.test
Tests export of configuration.
3. config_download_full_export in core/modules/config/config.admin.inc
Downloads a tarball of the site configuration.
4. FileDirectoryTest::testFileDirectoryTemp in core/modules/simpletest/tests/file.test
Ensure that the file_directory_temp() function always returns a value.
5. system_file_system_settings in core/modules/system/system.admin.inc
Form builder; Configure the site file handling.
6. system_requirements in core/modules/system/system.install
Implements hook_requirements().

So the last one shows it gets called during the installation. That's why on a system that was just installed you see the path on UI, but the config file is not aware of the path.

**Comment by jenlampton** [comment]
It looks like a few of these might need to be updated to use `config_get()`. I'm thinking `config_download_full_export()` and `ConfigurationUITest::testExport` for sure. Maybe also `FileDirectoryTest::testFileDirectoryTemp`?

**Comment by alanmels** [comment]
> It looks like a few of these might need to be updated to use `config_get()`. I'm thinking `config_download_full_export()` and `ConfigurationUITest::testExport` for sure. Maybe also `FileDirectoryTest::testFileDirectoryTemp`?

I agree: once we have the temporary directory set in the configuration file, there is no need to call `file_directory_temp()` at all. I could take a stab for another PR to incorporate all suggestions, but shouldn't we address this single issue first? Taking into account it is a bug, indeed!

**Comment by jenlampton** [comment]
I was thinking we should do it all at once because this fix adds a `config_set()` into a function that's called more than it should be. I dunno though, we may need a core committer's input. @herbdool @quicksketch @serundeputy?

**Comment by herbdool** [comment]
My feeling is that this isn't a bug. It's probably the same in d7 but I haven't checked.

**Comment by herbdool** [comment]
My feeling is that this isn't a bug. It's probably the same in d7 but I haven't checked.

**Comment by alanmels** [comment]
@herbdool Stay assured it is not the same on D7. Drupal 7 does not have configuration files.

**Comment by herbdool** [comment]
yes, @alanmels, I'm aware that D7 doesn't have configuration files. That's not what I meant.

Let's start off comparing the D7 `file_directory_temp` with Backdrop's `file_directory_temp`. From what I can tell, in D7 when `empty($temporary_directory)` is true it does some discovery or builds up a tmp directory from the public files directory and then it saves it to a variable. That save only happens once since every subsequent call will have a `$temporary_directory`. So that saving is what's not present in Backdrop.

It would be good to look at the `git blame` to see why it's not there. Is there a related d.org or Backdrop issue that gives a reason for its removal?

**Comment by alanmels** [comment]
@herbdool good, now you've looked at some code and saw by yourself D7 saves it variable, which is in turn picked up when a user goes to configuration page. So there is no discrepancy between what users sees on UI and configuration setting.

I don't know why and what was removed from Backdrop, however I believe what is important at that point is to quickly address the issue. Core should not have such bugs.

**Comment by indigoxela** [comment]
@alanmels this isn't a bug.

The fact that you didn't expect it to work that way doesn't mean, there's an urgent problem. It works fine.

I usually fully rely on the php.ini settings. Because of that I'm sure that:

1) The temporary directory exists on dev _and_ prod (I don't have to create/check).
2) That directory is writable by my php processes.

If that value gets set on install it means there needs to be some additional check when moving to a different server. So your proposed fix might actually make it worse.

**Comment by indigoxela** [comment]
@alanmels I've left a comment at your PR.
My recommendation is to set the item only once on install and not each and every time file_directory_temp() is called.

**Comment by alanmels** [comment]
> @alanmels this isn't a bug.
> 
> The fact that you didn't expect it to work that way doesn't mean, there's an urgent problem. It works fine.
> 
> I usually fully rely on the php.ini settings. Because of that I'm sure that:
> 
> 1. The temporary directory exists on dev _and_ prod (I don't have to create/check).
> 2. That directory is writable by my php processes.
> 
> If that value gets set on install it means there needs to be some additional check when moving to a different server. So your proposed fix might actually make it worse.

@indigoxela, thanks for the input. Your statement is absolutely true if it was about us, developers. However, what you are missing IMO is the fact that when a regular user have just installed Backdrop, thens/he sees on UI the temp directory, which is in fact missing in the config file. Then that user could easily try to enable another contributed or custom module which tries to use the `config_get('system.core', 'file_temporary_path')` method and all kind of problems start.

How don't you, guys, understand you can visually feed to anyone what has not been really set in the respective config file? What a user sees on UI is at least deceptive information, at most that info can cause further problems like data loss is sensitive information tried to save, but failed. This is a serious bug, guys, you just need to run some tests by yourselves to see that.

**Comment by alanmels** [comment]
> @alanmels I've left a comment at your PR.

Yep, and here it is:

> Setting the tmp dir shouldn't happen in function file_directory_temp() in file.inc, as that function is called several times, which means, the config item gets set again and again.
> 
> See: https://api.backdropcms.org/api/backdrop/core%21includes%21file.inc/function/file_directory_temp/1
> 
> It should rather go to core/modules/system/system.install to make sure, it only happens once on install.

And here is my answer to your comment:

> That's exactly why @jenlampton has put this comment https://github.com/backdrop/backdrop-issues/issues/4193#issuecomment-550064054 So we should call it only once - during the install - and all the other instances should replace with `config_get()`.
> 
> However, even if we didn't try to replace other calls to the function, I don't agree that it  
> 
> > gets set again and again
> 
> because there is a check in the very beginning of the module:
> 
> ```
> function file_directory_temp() {
>   $temporary_directory = config_get('system.core', 'file_temporary_path');
>   if (empty($temporary_directory)) {
> ```
> 
> In other words, the execution never proceeds if temp directory has already been put into the config file. So nothing terrible in case of multiple calls too.

**Comment by indigoxela** [comment]
@alanmels many thanks for further clarifying your position.

In the meantime - your PR needs work. :wink:

**Comment by alanmels** [comment]
> @alanmels many thanks for further clarifying your position.
> 
> In the meantime - your PR needs work. 😉

Happy to hear specific suggestions. @jenlampton did offer specific suggestion, I just had not time to dive into it. Likewise let me know what exactly should be changed in your opnion.

**Comment by alanmels** [comment]
@indigoxela, to keep discussions separately from the commit page, copying our last two comments:

Your last:

> > So nothing terrible in case of multiple calls too.
> 
> True, but there's no need to do it here. Running config_set once at install would be much cleaner. My recommendation is still to _not_ do this in file.inc but in the installer.

And my reaction to it:

> Well, I'll be more happy if core gets more optimized. The multiple checks have already been there. They are not caused by my PR.
> 
> My PR is trying to deal with specific problem that Backdrop UI is giving wrong information to user. But let's see how this goes - if core maintainers would be willing to introduce multiple changes instead one suggested.
> 
> Your idea to do checks on installer is quite logical. In fact, if the function gets called only once by installer, then there is no need for it in the module - the whole function needs to move to installer.

**Comment by indigoxela** [comment]
> Likewise let me know what exactly should be changed in your opinion.

Like my comment at your PR says: do not do this in file_directory_temp(), as that one is meant to be a utility function to "get the path of system-appropriate temporary directory".

To my opinion the right place would be the installer for:

1) new sites to set the value initially
2) existing sites to add the config value if it hasn't been set yet

But only once. And I hope, you agree on that. (Ah, you seem to :wink: )

**Comment by alanmels** [comment]
> > Likewise let me know what exactly should be changed in your opinion.
> 
> Like my comment at your PR says: do not do this in file_directory_temp(), as that one is meant to be a utility function to "get the path of system-appropriate temporary directory".
> 
> To my opinion the right place would be the installer for:
> 
> 1. new sites to set the value initially
> 2. existing sites to add the config value if it hasn't been set yet
> 
> But only once. And I hope, you agree on that. (Ah, you seem to 😉 )

Of course I agree with the logic, @indigoxela, it should save only once. But then still I have to disagree with:

> Like my comment at your PR says: do not do this in file_directory_temp(), as that one is meant to be a utility function to "get the path of system-appropriate temporary directory".

Because, while you are absolutely right stating the function is to get the path, you are missing that the function forget to save that path to config file. Do you see what I mean? Either all the calls to it should be removed and the function itself should be moved to the installer. If the function is to remain intact, then the suggested commit has to be taken in. 

See @herbdool also realized in https://github.com/backdrop/backdrop-issues/issues/4193#issuecomment-550089859 that the original code on D7 saves the config. So the save feature of the functions has to be re-instated in any case.

**Comment by alanmels** [comment]
Second stab takes different approach per @indigoxela's suggestion. He insists the temporary directory should be saved in the `system.install` file, so please review https://github.com/backdrop/backdrop/pull/2977 instead?

Should I close the first PR then?

**Comment by alanmels** [comment]
What @herbdool suggested on PR's page is really worth considering:

> > @herbdool: Putting it in system.install may be cleaner but perhaps this PR helps in case where someone manually deletes the variable so it can automatically recreate it.
> 
> @alanmels Interesting catch. There will be some folks who get dirty their hands deleting entries in the config file. Then we are back to the situation when UI shows a path that is not present in the config file. So my first PR might still be better than the second one.

**Comment by klonos** [comment]
I have not reviewed this in depth, but here is my understanding and feedback, from a UX point of view (I am making some assumptions - I could be wrong):

- There seem to be use cases where even though the value for the temp dir has not been set (in config), we set the default value of the field in the form.
- The above has the upside of setting a sensible default (for when the user saves the form)
- The downside is that, at the same time, it does not reflect reality; so not ideal.
- I think that it is the right move to first check if the config value has been set. If not, then set it. Once it's been set, we can read that, and set the default value for the field. The form then would reflect reality.

If I am missing something, please point it out.

**Comment by quicksketch** [comment]
I think when @indigoxela and @herbdool suggested we set the value in system.install, they were implying it should go in `system_install()` so it's there on new installs, rather than in `system_requirements()`. I agree with that approach too. So an outline to fix this:

- Set the config value in system_install(). It's worth noting that we already set `file_public_path` there. Why we're not also setting `file_temporary_path` isn't apparent to me.
- For existing sites, add an update hook `system_update_1070()` (currently), that also sets `file_temporary_path` just for good measure.

Some reasons why I don't think we should add config_set() at the bottom of `file_directory_temp()`:

- As @alanmels has suggested, confusion arises when trying to use `config_get('system.core', 'file_temporary_path')` and it's empty. Adding `config_set()` at the bottom of `file_directory_temp()` won't help in 100% of situations, because it has to be called at least once to set the value in config. The value should just be in config all the time.
- `file_directory_temp()` could be called any number of places, from it's documentation and the way it's used, the expectation is that it's only reading a value, not setting one. As far as I know, we don't have any places in core that _write_ config just by viewing a page. Config shouldn't be changed unless an explicit action is taken by an administrator, like saving a form or running update.php.

@indigoxela also noted that sometimes there are good reasons to have different config values on different environments, and the temporary directory is a good candidate for such a scenario. We don't have the ability to override config in settings.php, but IMO that would be the best way to handle per-environment config values, we have an issue for that at #2805.

**Comment by klonos** [comment]
I have tested the PR on my local, and confirmed that there is a `"file_temporary_path": "/tmp",` line added in `system.core.json` on a fresh install.

I have also manually applied the code changes on an existing site, which did not have `file_temporary_path` set in config. After running db update, the setting got populated.

I then finally tested this on another existing test site, after having saved the temp directory as `/123`. Running db update did not change that setting.

So WFM 👍 ...unfortunately, since I have contributed to the PR code, I cannot RTBC this. Someone else will need to do the code review.

**Comment by klonos** [comment]
PS: both php5 and php7 tests have failed with the same time/timezone random failure. Totally unrelated to the code changes.

**Comment by alanmels** [comment]
> PS: both php5 and php7 tests have failed with the same time/timezone random failure. Totally unrelated to the code changes.

In fact, all the checks passed after I closed and re-opened the https://github.com/backdrop/backdrop/pull/2988

**Comment by quicksketch** [comment]
Thanks @alanmels, that looks perfect! Merged https://github.com/backdrop/backdrop/pull/2988 into 1.x and 1.14.x.

Thanks also to @klonos, @indigoxela, @herbdool, @BWPanda, and @jenlampton for their feedback on this issue!
